File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/H._Подстроки/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(K), where K is length of distinct characters
    """
    sentence: str = sys.stdin.readline().rstrip()
    hashmap: dict[str, int] = dict()
    max_sub: int = 0
    left: int = 0
    for right in range(len(sentence)):
        hashmap[sentence[right]] = hashmap.get(sentence[right], 0) + 1
        if len(hashmap) != right - left + 1:
            hashmap[sentence[left]] -= 1
            if not hashmap[sentence[left]]:
                del hashmap[sentence[left]]
            left += 1
        max_sub = max(max_sub, right - left + 1)
    sys.stdout.write(str(max_sub) + '\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/H._Подстроки/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(K), where K is length of distinct characters
    */
    std::string sentence;
    std::cin >> sentence;
    std::unordered_map<char, int> hashmap;
    int maxSub = 0;
    int left = 0;
    for (int right = 0; right < sentence.size(); ++right) {
        ++hashmap[sentence[right]];
        if (hashmap.size() != right - left + 1) {
            --hashmap[sentence[left]];
            if (!hashmap[sentence[left]]) {
                hashmap.erase(sentence[left]);
            }
            ++left;
        }
        maxSub = std::max(maxSub, right - left + 1);
    }
    std::cout << maxSub << "\n";
}


int main() {
    solution();
}